### ISSUE TITLE
Support adding files to <input type="file"> elements

### ISSUE THREAD

**Issue by cigitia** [electron/electron#749]
It’s currently **not possible to programmatically upload files** using Electron. **Electron should support adding local files to `<input type="file">`** elements.

Electron primarily uses file-path strings to read files: the purpose of both the [dialog API](https://github.com/electron/electron/blob/master/docs/api/dialog.md) and the [File DOM object’s added API](https://github.com/electron/electron/blob/master/docs/api/file-object.md) is to return file-path strings, which presumably will then be read by some node.js module.

It would be wonderful, however, if electron also somehow supported adding local files to `<input type="file">` elements. For instance, a browser application that uses Electron may need to navigate to a webpage, then add a file chosen by the user to the webpage—without activating an OS-native file-choosing dialog. This might be because the application provides its own UI for choosing files, or it might be (as in my case) because the application’s purpose is to automate repetitive website forms that require the same files to be uploaded.

There are two ways I can think of doing this:
1. Making it possible both to set the `files` properties of `<input type="file">` elements to `FileList` objects (as NW.js does), to and construct `FileList` objects file-path strings (either arrays of strings, or `File` objects constructed from strings). This has the advantage of making it possible to directly reuse `FileList`s from drag and drop events, as well as other `<input type="file">` elements. `FileList`s need not be mutable; the programmer could construct new ones every time they need them.
   
   [Edit: This is also the solution that NW.js has done.]
2. Making it possible to set the `files` properties of `<input type="file">` elements to arrays of file-path strings. (This has the advantage of using strings directly, which may be simpler in most cases.)

NW.js, the other node–Chromium project, already supports setting the `files` property of `<input type="file">` elements, but that requires a preexisting `FileList` DOM object: <del>it too does not support creating new `FileList`s</del>. [Edit: This is actually incorrect. NW.js does also support creating new `FileList`s. See comment below.]

The Electron issue most related to this one is #454 (“Clicking on file input elements doesn't show file dialog”), but it is not the same, and this issue actually would determine whether I could use Electron for my own application.

In any case, Electron is amazing; cheers to all who have helped with it.

Related links:

* Electron issues and documentation
    * [Electron DOM `File` extensions](http://electron.atom.io/docs/api/file-object/)
    * #454 “Clicking on file `input` elements doesn't show file dialog”
    * #1374 “webview dialog event”
    * #3999 “Overriding `<webview>`’s file-upload dialog”
    * #5949 “Grabbing file path from `input=file`”
    * #6809 “Can `dialog#showOpenDialog` return `File` instances instead of paths?”
* Electron-using projects’ issues and hacks
    * segmentio/nightmare#235 “upload functionality removed?”
    * [jakubzitny/electron-file-hack](/jakubzitny/electron-file-hack/blob/master/renderer.js)
* NW.js issues and documentation
    * nwjs/nw.js#4174 “Overriding policies for `window`/`<webview>`/`<iframe>` file-upload, alert, confirmation, prompt dialogs”
    * nwjs/nw.js#5462 “Webview event for common dialogs”
    * [NW.js `Window`: `new-win-policy` event](/nwjs/nw.js/wiki/Window#user-content-new-win-policy)
    * [NW.js `File`-value functional test](/nwjs/nw.js/blob/master/tests/automatic_tests/filevalue/index.html)
* DOM documentation
    * [W3C File API spec, current draft](https://w3c.github.io/FileAPI/)
    * [MDN `File` reference](https://developer.mozilla.org/en-US/docs/Web/API/File)
    * [Stack Overflow: “How to instantiate a File object in JavaScript?”](https://stackoverflow.com/questions/8390855/how-to-instantiate-a-file-object-in-javascript)
* Browser-extension documentation
    * [Firefox legacy browser-extension API: “Using the DOM File API in chrome code”](https://developer.mozilla.org/en-US/docs/Extensions/Using_the_DOM_File_API_in_chrome_code)
* Apple Cocoa documentation
    * [Stack Overflow: “Cocoa webkit: how to get file upload / file system access in webkit”](http://stackoverflow.com/questions/5177640/cocoa-webkit-how-to-get-file-upload-file-system-access-in-webkit)
    * [Legacy WebKit Cocoa framework: `WebUIDelegate` class’s `webView:runOpenPanelForFileButtonWithResultListener:` method](//developer.apple.com/library/mac/documentation/Cocoa/Reference/WebKit/Protocols/WebUIDelegate_Protocol/index.html#//apple_ref/occ/instm/NSObject/webView:runOpenPanelForFileButtonWithResultListener:)
* Google Chrome DevTools debugging protocol
    * [DevTools `DOM`-domain operations, version 1.2](https://chromedevtools.github.io/devtools-protocol/1-2/DOM/), which does *not* support setting file-input files anymore

**Comment by cdroms** [comment]
Hello!
Has there been any progress on this issue since it was logged?

**Comment by fritx** [comment]
Hello!
Has there been any progress on this issue since it was logged?

**Comment by erikmellum** [comment]
This is a feature that would be incredibly useful. It's near impossible to programmatically upload a file in electron right now. It is allegedly trivial in NW.js as mentioned by citigia above! Please add!

**Comment by lordgreg** [comment]
Has there been any news about this subject lately?

**Comment by kit92** [comment]
Does anyone know where in the chromium blink JS engine I could find the handler for this? I would be willing to dive in the code and see if I could make the change to the methods to make this work. I'd just rather not have to read all the chromium code to find this one issue.

**Comment by svbatalov** [comment]
Hi.
We are making a browser automation tool (called The Bots Factory, or TBF) and one of our 'bots' required setting files in file input without user interaction. The trick that finally worked is to use chrome remote debugging, something like this:
```
// wv -- WebView
var wc = wv.getWebContents();
try {
  wc.debugger.attach("1.1");
} catch (err) {
  console.error("Debugger attach failed : ", err);
};

wc.debugger.sendCommand("DOM.getDocument", {}, function (err, res) {
  wc.debugger.sendCommand("DOM.querySelector", {
    nodeId: res.root.nodeId,
    selector: "#file"  // CSS selector of input[type=file] element
  }, function (err, res) {
    wc.debugger.sendCommand("DOM.setFileInputFiles", {
      nodeId: res.nodeId,
      files: ['/tmp/tst']  // Actual list of paths
    }, function (err, res) {
      wc.debugger.detach();
    });
  });
});
```
Note that the DevTools for the webview (in this case) must be closed for this to work.

Relevant pages:

- [Chrome remote debugging protocol 1.1](http://chromedevtools.github.io/debugger-protocol-viewer-1.1/DOM/)
- [WebContents debugger API](http://electron.atom.io/docs/v0.37.2/api/web-contents/#webcontentsdebugger)

**Comment by rosshinkley** [comment]
@svbatalov's example works, but it appears that the MIME type determination based on the registry/XDG is skipped (it looks like [`include_platform_types` might be false?](https://github.com/segmentio/nightmare/issues/787)).  Some spot-checking leads me to believe that the [hard coded](https://cs.chromium.org/chromium/src/net/base/mime_util.cc?sq=package:chromium&rcl=1475162463&l=64-112) MIME types will get determined properly, but anything else will fall to the default of `application/octet-stream`.

Is there a way to force the type check or explicitly set the type from the debugger?

**Comment by idododu** [comment]
+1

I am doing a upload app using electron and an opensource js upload library, and it works great in a webview.
I would like to add a feature: when open this app, it will continue to upload the file.
If I could set default files list for input file, then it will be easy to implement.

**Comment by idododu** [comment]
@svbatalov , 
1. should I put your codes in main process or render process?
2. do I have to use a webview tag in my html file?

**Comment by svbatalov** [comment]
AFAICT this approach should work as long as corresponding [webContents](http://electron.atom.io/docs/api/web-contents/#class-debugger) is available (mind the DevTools restriction). Take a look at the following [gist](https://gist.github.com/svbatalov/316da55cb0fb03e30eb1be1b49afd734) for simple demonstration (most of the code is identical to the above).

**Comment by idododu** [comment]
Tried your git, it works great!!!

detach debugger first to avoid 'another debugger has been attached' issue!
`try {
        if(wc.debugger.isAttached()) {
            wc.debugger.detach();
        }

        wc.debugger.attach("1.1");
    } catch (err) {
        console.error("Debugger attach failed : ", err);
    }`

**Comment by srkunze** [comment]
Would be great if there's native solution for it.

**Comment by ios122** [comment]
@ericbrumer 

> It's near impossible to programmatically upload a file in electron right now.

Maybe You hava find the way to do this,But I also want to share my idear.

Now, It is easy to programmatically upload a file in electron!!!

In your render process, you can get  the Whole cookies include session cookies:

```
  const {session} = require('electron').remote

  session.defaultSession.cookies.get({
    url: window.location.href
  }, (error, cookies) => {
    console.log(error, cookies)
  })
```

And then you can programmatically upload a file with some nodejs library:

```
const request = require("request")

request({
  url: `http://www.example.com/api2/upload`,
  method:"POST",
  formData: {
    upload_type: "zip",
    file: fs.createReadStream(pathToAZip)
  },
  headers:{
    "cookie": cookie // a string,not a array.Looks like:  "i18next=zh-CN;appType=0;"
  }
},function (err, response, body) {
  console.log(err,response,body)
})
```

**Comment by WaiSiuKei** [comment]
If there is node integration in the renderer process of your app, you can read the file as stream and then convert it to Blob.

**Comment by samsamm777** [comment]
Anything further on this? Im really struggling to get this working. I just need to create a File object from a path string.

**Comment by idododu** [comment]
@samsamm777 check svbatalov and my workaround above

**Comment by radreamer** [comment]
Hi guys, i faced with the problem when trying to upload folder. I have the input:
`<input type="file" webkitdirectory="true" onChange={e => { console.log(e.target.files) }} />`
When i select folder with many files, i got FileList array with only one File instance – that directory, that i choose earlier. When i try it in the browser – all works as expected, e.g. http://jsfiddle.net/addyo/Ha979/

**Comment by jarry** [comment]
I have met same issues with @radarhere . I got the FileList that's length 1 in Electron when using <input directory>. In Chrome, I can get all files.

**Comment by sqqihao** [comment]
> Hi.
> We are making a browser automation tool (called The Bots Factory, or TBF) and one of our 'bots' required setting files in file input without user interaction. The trick that finally worked is to use chrome remote debugging, something like this:
> 
> ```
> // wv -- WebView
> var wc = wv.getWebContents();
> try {
>   wc.debugger.attach("1.1");
> } catch (err) {
>   console.error("Debugger attach failed : ", err);
> };
> 
> wc.debugger.sendCommand("DOM.getDocument", {}, function (err, res) {
>   wc.debugger.sendCommand("DOM.querySelector", {
>     nodeId: res.root.nodeId,
>     selector: "#file"  // CSS selector of input[type=file] element
>   }, function (err, res) {
>     wc.debugger.sendCommand("DOM.setFileInputFiles", {
>       nodeId: res.nodeId,
>       files: ['/tmp/tst']  // Actual list of paths
>     }, function (err, res) {
>       wc.debugger.detach();
>     });
>   });
> });
> ```
> Note that the DevTools for the webview (in this case) must be closed for this to work.
> 
> Relevant pages:
> 
> * [Chrome remote debugging protocol 1.1](http://chromedevtools.github.io/debugger-protocol-viewer-1.1/DOM/)
> * [WebContents debugger API](http://electron.atom.io/docs/v0.37.2/api/web-contents/#webcontentsdebugger)

nice bro

**Comment by Freshdouble** [comment]
Are there any update on this?

I want to automate the octopart bom tool lookup. So i want to upload a csv file from my computer to the input form on their website and show the result page in a electron window.

I've tried the webview method that was posted above but the commands wc.debugger.sendCommand("DOM.getDocument", {}, function (err, res) didn't work and it never got to the next line.

**Comment by hedgehog90** [comment]
I would also like to see this implemented.

**Comment by rr13k** [comment]
I've tried the webview method that was posted above but the commands wc.debugger.sendCommand("DOM.getDocument", {}, function (err, res) didn't work and it never got to the next line.
debugger.attach succeeded however.

Is there anyone  I'm still looking at it 。
It's what I need ，Can anyone push it

**Comment by SomaticIT** [comment]
@rr13k FYI, `wc.debugger.sendCommand` is now a `Promise`.

So you should adapt the example above like so:

```javascript
async function setFileInput(wc, selector, files) {
  try {
    wc.debugger.attach("1.1");

    const { root } = await wc.debugger.sendCommand("DOM.getDocument", {});
    const { nodeId } = await wc.debugger.sendCommand("DOM.querySelector", { nodeId: root.nodeId, selector });

    await wc.debugger.sendCommand("DOM.setFileInputFiles", { nodeId, files });
  }
  finally {
    wc.debugger.detach();
  }
}

// then somewhere in your code:
await setFileInput(win.webContents, "#file", ["/tmp/test.txt"]);
```

**Comment by ngustavo** [comment]
Any news on this topic?

**Comment by croqaz** [comment]
Ended up here, because I have the same issue.
This can be hacked by making a button + input and implementing the "click" events manually, but it would be nice if this wasn't needed.

**Comment by codebytere** [comment]
This works as expected in Electron and looks like it has for some time -

```js 
// main.js
const { app, BrowserWindow } = require('electron')
const path = require('path')

function createWindow () {
  const mainWindow = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      preload: path.join(__dirname, 'preload.js')
    }
  })

  mainWindow.loadFile('index.html')
}

app.whenReady().then(() => {
  createWindow()

  app.on('activate', function () {
    if (BrowserWindow.getAllWindows().length === 0) createWindow()
  })
})

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') app.quit()
})
```

```html
// index.html
<div class="container">
  <div class="page-header">
    <h1>Demo <small>Drag and drop a directory on file input with <em>webkitdirectory</em></small></h1>
  </div>

  <div class="row">
    <input type="file" multiple webkitdirectory id="fileURL"/>
    <ul id="fileOutput"></ul>
  </div>
</div>
<script src="./renderer.js"></script>
```

```js
// renderer.js
(function () {
    var files,
        file,
        extension,
        input = document.getElementById("fileURL"),
        output = document.getElementById("fileOutput");

    input.addEventListener("change", function (e) {
        files = e.target.files;
        output.innerHTML = "";

        for (var i = 0, len = files.length; i < len; i++) {
            file = files[i];
            extension = file.name.split(".").pop();
            output.innerHTML += "<li class='type-" + extension + "'>" + file.name + "</li>";
        }
    }, false);
})();
```

Works the same in both Chrome and Electron.

**Comment by bughit** [comment]
@codebytere 

This request is about programmatically adding files to be uploaded to a file input. Your example does not show that, it shows reading file names from a file input.

**Comment by samsamm777** [comment]
@codebytere This is not a solution to the problem.

**Comment by bughit** [comment]
@zcbenz @codebytere Can someone please review the reason for this closure. The posted example:

https://github.com/electron/electron/issues/749#issuecomment-2254559878

does not address the problem:

> This request is about programmatically adding files to be uploaded to a file input. Your example does not show that, it shows reading file names from a file input.

**Comment by bughit** [comment]
@ckerr @MarshallOfSound @deepak1556 

Can someone from Electron please explain how [this example](https://github.com/electron/electron/issues/749#issuecomment-2254559878) shows this issue as "completed"?

**Comment by LqdBcnAtWork** [comment]
I'd also like to know why this is closed?

**Comment by thomasbachem** [comment]
Yes, please reopen this issue!

**Comment by bxclib2** [comment]
Nowadays, I see a lot of repo maintainers like this. When someone raise an issue, they gave some answer not related to the question, and then directly close. 

@codebytere I really cannot understand, why you gave this answer. Also, I don't understand why you close this useful issue.

**Comment by Veiintc** [comment]
any update?
